runtime/compiler/z/codegen/J9S390PrivateLinkage.cpp

```diff
@@ -3051,6 +3051,97 @@ void TR::J9S390JNILinkage::acquireVMAccessMask(TR::Node * callNode, TR::Register
+                                                    TR::Register * methodMetaDataVirtualRegister,
+                                                    TR::Register * tempReg1)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)fe();
+   TR::CodeGenerator* cg = self()->cg();
+   TR::Compilation* comp = self()->comp();
+
+   // Store 1 into vmthread->inNative
+   generateRIInstruction(cg, TR::InstOpCode::getLoadHalfWordImmOpCode(), callNode, tempReg1, 1);
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), callNode, tempReg1,
+                         generateS390MemoryReference(methodMetaDataVirtualRegister, offsetof(J9VMThread, inNative), cg));
+
+#if !defined(J9VM_INTERP_ATOMIC_FREE_JNI_USES_FLUSH)
+   // The asymmetric algorithm does not need this serialization.
+   TR::InstOpCode::S390BranchCondition cond = TR::InstOpCode::COND_MASK14;
+   if (!cg->getS390ProcessorInfo()->supportsArch(TR_S390ProcessorInfo::TR_z196))
+      cond = TR::InstOpCode::COND_MASK15;
+
+   TR::RealRegister *gpr0 = cg->machine()->getS390RealRegister(TR::RealRegister::GPR0);
+   new (cg->trHeapMemory()) TR::S390RegInstruction(TR::InstOpCode::BCR, callNode, cond, gpr0, cg);
+#endif
+
+   // Compare vmthread public flag with J9_PUBLIC_FLAGS_VM_ACCESS
+   generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), callNode, tempReg1,
+                         generateS390MemoryReference(methodMetaDataVirtualRegister, fej9->thisThreadGetPublicFlagsOffset(), cg));
+
+   TR::LabelSymbol * longReleaseSnippetLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol * longReleaseRestartLabel = generateLabelSymbol(cg);
+
+   TR::InstOpCode::Mnemonic cmpImmAndJumpOp = TR::Compiler->target.is64Bit() ? TR::InstOpCode::CGIJ : TR::InstOpCode::CIJ;
+   generateRIEInstruction(cg, cmpImmAndJumpOp, callNode, tempReg1, J9_PUBLIC_FLAGS_VM_ACCESS, longReleaseSnippetLabel, TR::InstOpCode::COND_BNE);
+
+   cg->addSnippet(new (self()->trHeapMemory()) TR::S390HelperCallSnippet(cg,
+                                                                         callNode, longReleaseSnippetLabel,
+                                                                         comp->getSymRefTab()->findOrCreateAcquireVMAccessSymbolRef(comp->getJittedMethodSymbol()),
+                                                                         longReleaseRestartLabel));
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, callNode, longReleaseRestartLabel);
+   }
+
+void
+TR::J9S390JNILinkage::acquireVMAccessMaskAtomicFree(TR::Node * callNode,
+                                                    TR::Register * methodMetaDataVirtualRegister,
+                                                    TR::Register * tempReg1)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)fe();
+   TR::CodeGenerator* cg = self()->cg();
+   TR::Compilation* comp = self()->comp();
+
+   // Zero vmthread->inNative, which is a UDATA field
+   uint32_t pointerSize = TR::Compiler->target.is64Bit() ? 7 : 3;
```

Comment:
I believe there is a helper for this either in `ObjectModel` or the base `CodeGenerator` class (for getting size of pointer). We should use that rather than magic constants.

Reply:
Changed to `sizeof(UDATA)-1`.

Reply:
I think what you want is `TR::Compiler->om.sizeofReferenceAddress() - 1`.

Reply:
OK